tests/robustness/multiclient_test/framework/harness.go

```diff
@@ -0,0 +1,241 @@
+	err = th.Cleanup(ctx)
+	if err != nil {
+		log.Printf("Error cleaning up the engine: %s\n", err.Error())
+		os.Exit(2)
+	}
+
+	if th.skipTest {
+		os.Exit(0)
+	}
+
+	os.Exit(1)
+}
+
+func (th *TestHarness) makeBaseDir() bool {
+	baseDir, err := ioutil.TempDir("", "engine-data-")
+	if err != nil {
+		log.Println("Error creating temp dir:", err)
+		return false
+	}
+
+	th.baseDirPath = baseDir
+
+	return true
+}
+
+func (th *TestHarness) getFileWriter() bool {
+	fw, err := NewMultiClientFileWriter(
+		func() (FileWriter, error) { return fiofilewriter.New() },
+	)
+	if err != nil {
+		if errors.Is(err, fio.ErrEnvNotSet) {
+			log.Println("Skipping robustness tests because FIO environment is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating fio FileWriter:", err)
+		}
+
+		return false
+	}
+
+	th.fileWriter = fw
+
+	return true
+}
+
+func (th *TestHarness) getSnapshotter() bool {
+	server, err := snapmeta.NewSnapshotter(th.baseDirPath)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping robustness tests because KOPIA_EXE is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating kopia Snapshotter:", err)
+		}
+
+		return false
+	}
+
+	newClientFn := func(baseDirPath string) (ClientSnapshotter, error) {
+		return snapmeta.NewSnapshotter(th.baseDirPath)
+	}
+
+	s, err := NewMultiClientSnapshotter(th.baseDirPath, newClientFn, server)
+	if err != nil {
+		log.Println("Error creating multiclient kopia Snapshotter:", err)
+		return false
+	}
+
+	th.snapshotter = s
+
+	if err = s.ConnectOrCreateRepo(th.dataRepoPath); err != nil {
+		log.Println("Error initializing kopia Snapshotter:", err)
+		return false
+	}
+
+	return true
+}
+
+func (th *TestHarness) getPersister() bool {
+	kp, err := snapmeta.NewPersister(th.baseDirPath)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping robustness tests because KOPIA_EXE is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating kopia Persister:", err)
+		}
+
+		return false
+	}
+
+	th.persister = kp
+
+	if err = kp.ConnectOrCreateRepo(th.metaRepoPath); err != nil {
+		log.Println("Error initializing kopia Persister:", err)
+		return false
+	}
+
+	return true
+}
+
+func (th *TestHarness) getEngine(ctx context.Context) bool {
+	args := &engine.Args{
+		MetaStore:        th.persister,
+		TestRepo:         th.snapshotter,
+		FileWriter:       th.fileWriter,
+		WorkingDir:       th.baseDirPath,
+		SyncRepositories: true,
+	}
+
+	eng, err := engine.New(args) // nolint:govet
+	if err != nil {
+		log.Println("Error on engine creation:", err)
+		return false
+	}
+
+	// Initialize the engine, connecting it to the repositories.
+	// Note that th.engine is not yet set so that metadata will not be
+	// flushed on cleanup in case there was an issue while loading.
+	err = eng.Init(ctx)
+	if err != nil {
+		log.Println("Error initializing engine for S3:", err)
```

Comment:
Why assume S3?  Wouldn't that depend on the configuration parameters?

Reply:
This is a cut-and-paste error. Thank you for catching it, will fix.